### PR TITLE
[ADD] marketing_card: create dynamic shareable card urls

### PR DESCRIPTION
**Pull request by reth-odoo**
Word-of-mouth is a great way to get high-quality leads for cheap.
This addon allows users to create templates for partners to share a link
with their a cover image tailored to them on social media.

A good use case is to allow presenters to advertize their track
during an event, each of them can advertize their own track and the
event in a way that's consistent accross all presenters of the event.

This also allows for rewarding partners who do share the event by
giving them any link once they share the event, to a coupon for example.

----------------------------------

### Open Graph

To get social networks to render the image, we use opengraph with a redirect url.
When a social network bot is detected on the redirection endpoint, the endpoint
serves a minimal html page containing only open graph data. When a regular user
then clicks the link they'll be redirected normally to a url of our choosing.

The list of network bot user agents needs to be kept up to date for this feature
to work properly, and the feature will only work for listed social networks.

### Rendering

The rendering is in two steps:
- ir.qweb to get an html web page
- wkhtmltoimage to actually render the image to a jpeg

wkhtmltoimage works very similarly to wkhtmltopdf, instanciating a headless
webkit engine to render the page. For our purposes the html body should be
self-sufficient, meaning wkhtmltopdf may not execute scripts or fetch data.

### Card Campaign

A card campaign comprises:

- A template, defining the general design/layout of the card in html
- Elements, modifiable by the user to either set static or dynamic values
- A target model, defining what model the dynamic values are picked from
- A preview record, allowing users to preview the card with specific values
- A send button, allowing users to send the card by mail using a domain

Each campaign tracks cards sent, cards opened, cards shared and how many clicks
these cards from social media globally.

The campaign elements simply hold values to be passed to ir.qweb at the time
of rendering. Either a static value (text/uploaded image), or a field path.
The field path is validated based on hard-coded values on each model.
Other field paths may only be set by system administrators, as they may
pose security concerns. **All paths are fetched as sudo**.

Sharing the card by other means than the built-in composer can be done
by just calling `_get_preview_url_from_res_id` on a specific card campaign
for any automated action and/or mail template. The card will be auto-generated
no matter what. Although it is recommended to create it uppon sending
for tracking purposes.

### Card Templates

Templates define the layout of the card in html as well as some
default color and background settings. For the base data the bodies
simply call views which makes it easier to reuse CSS classes and
qweb variables.

### Cards

Cards cache the resulting card image for any recipient so as to prevent compute
spamming, as rendering is relatively expensive. For the same reason card URLs
are hmac-signed so users cannot generate cards for all recipients at once
by just guessing url parameters.

One a user has shared the card, they'll get a message on the preview page
which can be used for them to get small rewards. The message is the same
for every participant, and the "share" detection mechanism is not totally
reliable in that it will lead to false positives. Therefore the rewards
should not be too consequential.

To avoid massive unused data, the cards are cleared of their images every day.
And (re-)computed on demand.

----------------------------------

task-3495279

### CODE REVIEW

**Comment by reth-odoo** [comment]
@tde-banana-odoo 100 lines of js later, should be good.

python-wise I now process the URLs in the composer, and hardcoded "/cards/" in the blacklisted urls for massmailing

wysiwyg-wise
Now it will pick a card if any exists, and switch to the default card whenever the model is changed. Unless no card exists for the model, then it behaves as before for now.

File: addons/mass_mailing_marketing_card/views/snippets/snippets.xml [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    <xpath expr="div[@data-js='ReplaceMedia']" position="attributes">
+        <attribute name="data-exclude" add=".s_call_to_share_card img" separator=", "/>
+    </xpath>
+
+    <!--Alignment doesn't work well for images inside anchors-->
+    <!--TODO fix for all anchored images-->
+    <xpath expr="div[.//we-select[@id='o_we_image_alignment_option']]" position="attributes">
+        <attribute name="data-exclude" add=".s_call_to_share_card img" separator=", "/>
+    </xpath>
+    <xpath expr="div[.//we-select[@id='o_we_image_alignment_option']]" position="after">
```
Comment (awa-odoo) [inline addons/mass_mailing_marketing_card/views/snippets/snippets.xml:38]: I don't understand the difference between doing this with a xpath or putting the option directly inside the MarketingCard snippet here below?

EDIT oh, I think I get it, it's to have those options in the appropriate sections.

File: addons/mass_mailing_marketing_card/models/card_card.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+from collections import defaultdict
+
+from odoo import models, osv
+
+
+class CardCard(models.Model):
+    _inherit = 'card.card'
+
+    def _create_if_not_exists(self, card_values_all):
```
Comment (awa-odoo) [inline addons/mass_mailing_marketing_card/models/card_card.py:9]: Don't really get why this is a separate method, seems highly specific to the mass mailing processing, not sure how anyone will re-use this.

File: addons/mass_mailing_marketing_card/models/card_card.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        """Create new cards based on precommit values."""
+        if not card_values_all:
+            return
+
+        # only keep the last one
+        values_from_campaign_res_pair = dict()
+        for values in card_values_all:
+            values_from_campaign_res_pair[(values['campaign_id'], values['res_id'])] = values
```
Comment (awa-odoo) [inline addons/mass_mailing_marketing_card/models/card_card.py:17]: Correct me if I'm wrong here, but it seems like a caller has a list of tuple, transforms it as a dict, sends it here, to then re-transform it as a list of tuple... ?

File: addons/mass_mailing_marketing_card/models/card_card.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        existing_res_ids_from_campaign_id = {
+            campaign.id: set(res_ids)
+            for campaign, res_ids in existing_res_ids_from_campaign_id
+        }
+
+        # create missing
+        self.env['card.card'].create([
+            values for (campaign_id, card_id), values in values_from_campaign_res_pair.items()
+            if res_id not in existing_res_ids_from_campaign_id.get(campaign_id, ())
+        ])
```
Comment (awa-odoo) [inline addons/mass_mailing_marketing_card/models/card_card.py:42]: This seems like a lot of added complexity.

What are the odds that a mailing would contain multiple card campaigns?
Could we just settle for simplicity, iterate on campaigns and call the pre-existing "_get_or_create_cards_from_res_ids" that already contain a similar logic?

**Review by awa-odoo — COMMENTED** [review]
Some wild thoughts on this, trying to iterate nicely.

(Note: Did not bother too much with cosmetic comments given the short time frame).

**Comment by awa-odoo** [comment]
@reth-odoo While I'm thinking about it, runbot doesn't seem happy for ci/style, and also because you add invisible fields inside views, which is now forbidden apparently.

Would be nice to fix those when you have some time on your quest to make the build green.

**Comment by reth-odoo** [comment]
Already on it ^^
Seems it stopped whining about wkhtmltoimage so we should be good once it's green

File: addons/mass_mailing_marketing_card/views/snippets/snippets.xml (hunk 1 above)
Comment (reth-odoo) [reply addons/mass_mailing_marketing_card/views/snippets/snippets.xml:38]: Yes I have to add "data-exclude" for elements inside the card so people can't modify them "normally" and break the whole layout (less than they can now at least).

File: addons/mass_mailing_marketing_card/wizards/mail_compose_message.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    @api.model
+    def _process_generic_card_url_body(self, res_id_body_pairs):
+        """Update the bodies with the specific card url for that res_id and create a card.
+
+        :example: (1, "/cards/9/preview") -> (1, "/cards/9/1/abchashtoken/preview") + new card as side-effect
+        :param list[tuple[int, str]] res_id_body_pairs:
+        :return: processed bodies in the order they were received
+        :rtype: list[str]
+        """
+        campaign_tokens = lazymapping(
```
[the remark this answers is not in the document]
Comment (reth-odoo) [reply addons/mass_mailing_marketing_card/wizards/mail_compose_message.py:23]: This is essentially a defaultdict that can use the key as argument for the "default" function.
`lazy` just evaluates lazily on accessing a variable.

Indeed I've made the whole create flow simpler as suggested.

**Comment by reth-odoo** [comment]
Moving back to the original PR as we know we want mass mailing now.

**Pull request closed without merging**